Funcky.Test/Extensions/EnumerableExtensions/MaterializeTest.cs

```diff
@@ -39,7 +40,77 @@ public void MaterializeWithMaterializationReturnsCorrectCollectionWhenEnumerate(
             Assert.IsType<HashSet<string>>(list.Materialize(ToHashSet));
         }
 
+        [Fact]
+        public void MaterializeDoesNotEnumerateCollectionWhichImplementsICollectionOnly()
+        {
+            var list = new FailOnEnumerateCollection<int>(Count: 10);
+            _ = list.Materialize();
+        }
+
+        [Fact]
+        public void MaterializedICollectionHasCorrectCount()
+        {
+            const int count = 10;
+            var list = new FailOnEnumerateCollection<int>(Count: count);
+            Assert.Equal(count, list.Materialize().Count);
+        }
+
+        [Fact]
+        public void MaterializingAnICollectionReturnsAnICollection()
+        {
+            var collection = new FailOnEnumerateCollection<int>(Count: 0);
+            Assert.IsAssignableFrom<ICollection<int>>(collection.Materialize());
+        }
+
+        [Fact]
+        public void MaterializingAnIListReturnsAnIList()
+        {
+            var collection = new FailOnEnumerateList<int>(Count: 0);
+            Assert.IsAssignableFrom<IList<int>>(collection.Materialize());
+        }
+
+        [Fact]
+        public void MaterializingAnIListReturnsAnIReadOnlyList()
+        {
+            var collection = new FailOnEnumerateList<int>(Count: 0);
+            Assert.IsAssignableFrom<IReadOnlyList<int>>(collection.Materialize());
+        }
+
         private static HashSet<string> ToHashSet(IEnumerable<string> s)
             => s.ToHashSet();
+
+        private record FailOnEnumerateCollection<T>(int Count) : ICollection<T>
```

Comment:
We already have a few Fail* classes, these two should also go into separate files

Reply:
b814f344db84bcd0fc2ca40f37380a832c8c0cc0